core/src/main/java/org/apache/iceberg/V2Metadata.java

```diff
@@ -0,0 +1,321 @@
+    }
+
+    @Override
+    public boolean hasAddedFiles() {
+      return wrapped.hasAddedFiles();
+    }
+
+    @Override
+    public Integer addedFilesCount() {
+      return wrapped.addedFilesCount();
+    }
+
+    @Override
+    public Long addedRowsCount() {
+      return wrapped.addedRowsCount();
+    }
+
+    @Override
+    public boolean hasExistingFiles() {
+      return wrapped.hasExistingFiles();
+    }
+
+    @Override
+    public Integer existingFilesCount() {
+      return wrapped.existingFilesCount();
+    }
+
+    @Override
+    public Long existingRowsCount() {
+      return wrapped.existingRowsCount();
+    }
+
+    @Override
+    public boolean hasDeletedFiles() {
+      return wrapped.hasDeletedFiles();
+    }
+
+    @Override
+    public Integer deletedFilesCount() {
+      return wrapped.deletedFilesCount();
+    }
+
+    @Override
+    public Long deletedRowsCount() {
+      return wrapped.deletedRowsCount();
+    }
+
+    @Override
+    public List<PartitionFieldSummary> partitions() {
+      return wrapped.partitions();
+    }
+
+    @Override
+    public ManifestFile copy() {
+      return wrapped.copy();
+    }
+  }
+
+  static Schema entrySchema(Types.StructType partitionType) {
+    return wrapFileSchema(DataFile.getType(partitionType));
+  }
+
+  static Schema wrapFileSchema(Types.StructType fileSchema) {
+    // this is used to build projection schemas
+    return new Schema(
+        ManifestEntry.STATUS, ManifestEntry.SNAPSHOT_ID,
+        required(ManifestEntry.DATA_FILE_ID, "data_file", fileSchema));
+  }
+
+  static class IndexedManifestEntry implements ManifestEntry, IndexedRecord {
+    private final org.apache.avro.Schema avroSchema;
+    private final long snapshotId;
+    private final V1Metadata.IndexedDataFile fileWrapper;
+    private ManifestEntry wrapped = null;
+
+    IndexedManifestEntry(long snapshotId, Types.StructType partitionType) {
+      this.avroSchema = AvroSchemaUtil.convert(entrySchema(partitionType), "manifest_entry");
+      this.snapshotId = snapshotId;
+      // TODO: when v2 data files differ from v1, this should use a v2 wrapper
+      this.fileWrapper = new V1Metadata.IndexedDataFile(avroSchema.getField("data_file").schema());
+    }
+
+    public IndexedManifestEntry wrap(ManifestEntry entry) {
+      this.wrapped = entry;
+      return this;
+    }
+
+    @Override
+    public org.apache.avro.Schema getSchema() {
+      return avroSchema;
+    }
+
+    @Override
+    public void put(int i, Object v) {
+      throw new UnsupportedOperationException("Cannot read using IndexedManifestEntry");
+    }
+
+    @Override
+    public Object get(int i) {
+      switch (i) {
+        case 0:
+          return wrapped.status().id();
+        case 1:
+          return wrapped.snapshotId();
+        case 2:
+          if (wrapped.sequenceNumber() == null) {
+            Preconditions.checkState(wrapped.snapshotId() == null || snapshotId == wrapped.snapshotId(),
```

Comment:
I think this also needs some detail or document about why it happens.

Reply:
I added a comment to explain here as well.